docs/misc/expression.md

```diff
@@ -1,68 +1,68 @@
 
-其中判断表达式等价的部分使用了拉格朗日插值法等数学工具，在此暂不进行展开。
+The part to check the equivalence of expressions uses mathematical tools such as Lagrangian interpolation, and will not be discussed here.
 
-一般的思路分为两种，一种递归一种非递归。
+There are two common methods: one is recursive and the other is non-recursive.
 
-## 递归
+## Recursive
 
-递归的方法是把表达式拆分成如图所示的表达式树，然后在树结构上自底向上进行运算。![](./images/bet.png)
+The recursive method is to split the expression into an expression tree as shown in the figure, and then perform operations on the tree from bottom to top. ![](./images/bet.png)
 
-表达式树上进行 [树的遍历](../graph/tree-basic.md#_10) 可以得到不同类型的表达式
+[Tree traversal](../graph/tree-basic.md#_10) on expression tree can get different types of expressions:
 
-- 前序遍历对应前缀表达式（波兰式）
-- 中序遍历对应中缀表达式
-- 后序遍历对应后缀表达式（逆波兰式）
+- Preorder traversal corresponding to prefix expressions (Polish Notation)
+- Middle order traversal corresponds to infix expression
+- Post-order traversal corresponding to postfix expressions (Reverse Polish Notation)
 
-## 非递归
+## Non-recursive
 
-非递归的方法是定义两个 [栈](../ds/stack.md) 来分别存储运算符和运算数。每当遇到一个数直接放进数的栈；每当遇到一个操作符时，要查找之前运算符栈中的元素，按照预先定义好的优先级来进行适当的弹出操作（弹出的同时求出对应的子表达式的值）。
+The non-recursive method is to define two [stacks](../ds/stack.md) to store operators and operands respectively. Whenever a number is encountered, it is directly pushed into the number stack; whenever an operator is encountered, the element in the previous operator stack must be searched, and the appropriate pop operation will be performed according to the pre-defined priority(The value of the corresponding sub-expression).
 
-我们要知道：算术表达式分为三种，分别是前缀表达式、中缀表达式、后缀表达式。其中，中缀表达式是我们日常生活中最常用的表达式；后缀表达式是计算机最容易理解的表达式。为什么说后缀表达式最容易被计算机理解呢？因为后缀表达式不需要括号表示，它的运算顺序是唯一确定的。举个例子：在后缀表达式 $3 2 * 1 -$ 中，首先计算 $3 \times 2 = 6$ （使用最后一个运算符，即栈顶运算符），然后计算 $6 - 1 = 5$ 。可以看到：对于一个后缀表达式，只需要 **维护一个数字栈，每次遇到一个运算符，就取出两个栈顶元素，将运算结果重新压入栈中** 。最后，栈中唯一一个元素就是改后缀表达式的运算结果时间复杂度 $O(n)$ 。
+What we need to know: the arithmetic expressions have three types, namely prefix expressions, infix expressions, and postfix expressions. Among them, infix expressions are the most commonly used expressions in our daily lives; Postfix expressions are the easiest expressions for computers to understand. Why is it that postfix expressions are the easiest to be understood by computers? Because the suffix expression does not need parentheses, and its order of operations is uniquely determined. For example: In the postfix expression $3 2 * 1 -$ , first we calculate $3 \times 2 = 6$ (using the last operator, that is, the operator on the top of the stack), and then calculate $6 - 1 = 5$ . It can be seen that for a suffix expression, we only need **to maintain a number stack. Each time an operator is encountered, the top elements from two stacks are popped out, and the result of the operation is pushed back into the stack**. Finally, the only element in the stack is the operation result of changing the suffix expression. The time complexity is $O(n)$ .
```

Comment:
```suggestion
What we need to know: the arithmetic expressions have three types, namely prefix expressions, infix expressions, and postfix expressions. Among them, infix expressions are the most commonly used expressions in our daily lives; Postfix expressions are the easiest expressions for computers to understand. The reason is that suffix expression does not need parentheses, and its order of operations is uniquely determined. For example: In the postfix expression $3~2~*~1~-$ , first we calculate $3 \times 2 = 6$ (using the last operator, that is, the operator on the top of the stack), and then calculate $6 - 1 = 5$ . It can be seen that for a suffix expression, we only need **to maintain a number stack. Each time an operator is encountered, the top elements from two stacks are popped out, and the result of the operation is pushed back into the stack**. Finally, the only element in the stack is the operation result of the suffix expression. The time complexity is $O(n)$ .
```
原文 typo + concise